File: apps/admin/src/auth/firebase-gateway.ts
"use client";

import { FirebaseError } from "firebase/app";
import {
  onAuthStateChanged,
  sendEmailVerification,
  sendPasswordResetEmail,
  signInWithEmailAndPassword,
  signOut,
  type User,
} from "firebase/auth";

import { getFirebaseClientAuth } from "@/firebase/client";

export type AuthSession = {
  email: string | null;
  emailVerified: boolean;
  getIdToken: () => Promise<string>;
  uid: string;
};

export class ClientAuthError extends Error {
  constructor(
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = "ClientAuthError";
  }
}

export interface AuthGateway {
  getIdToken(): Promise<string | undefined>;
  observe(listener: (session: AuthSession | null) => void): () => void;
  refreshSession(): Promise<AuthSession>;
  sendEmailVerification(): Promise<void>;
  sendPasswordResetEmail(email: string): Promise<void>;
  signIn(email: string, password: string): Promise<AuthSession>;
  signOut(): Promise<void>;
}

const toSession = (user: User): AuthSession => ({
  email: user.email,
  emailVerified: user.emailVerified,
  getIdToken: () => user.getIdToken(),
  uid: user.uid,
});

function translate(error: unknown): ClientAuthError {
  if (error instanceof FirebaseError) return new ClientAuthError(error.code, error.message);
  if (error instanceof Error) return new ClientAuthError("auth/client-error", error.message);
  return new ClientAuthError("auth/client-error", "Firebase authentication failed");
}

export class FirebaseAuthGateway implements AuthGateway {
  async getIdToken(): Promise<string | undefined> {
    return getFirebaseClientAuth().currentUser?.getIdToken();
  }

  observe(listener: (session: AuthSession | null) => void): () => void {
    try {
      return onAuthStateChanged(getFirebaseClientAuth(), (user) =>
        listener(user && toSession(user)),
      );
    } catch (error) {
      throw translate(error);
    }
  }

  async signIn(email: string, password: string): Promise<AuthSession> {
    try {
      const credential = await signInWithEmailAndPassword(getFirebaseClientAuth(), email, password);
      return toSession(credential.user);
    } catch (error) {
      throw translate(error);
    }
  }

  async refreshSession(): Promise<AuthSession> {
    try {
      const user = getFirebaseClientAuth().currentUser;
      if (!user) throw new ClientAuthError("auth/no-current-user", "No Firebase session");
      await user.reload();
      await user.getIdToken(true);
      return toSession(user);
    } catch (error) {
      if (error instanceof ClientAuthError) throw error;
      throw translate(error);
    }
  }

  async sendEmailVerification(): Promise<void> {
    try {
      const user = getFirebaseClientAuth().currentUser;
      if (!user) throw new ClientAuthError("auth/no-current-user", "No Firebase session");
      await sendEmailVerification(user);
    } catch (error) {
      if (error instanceof ClientAuthError) throw error;
      throw translate(error);
    }
  }

  async sendPasswordResetEmail(email: string): Promise<void> {
    try {
      const actionUrl = process.env.NEXT_PUBLIC_AUTH_ACTION_URL?.trim();
      await sendPasswordResetEmail(
        getFirebaseClientAuth(),
        email,
        actionUrl ? { handleCodeInApp: false, url: actionUrl } : undefined,
      );
    } catch (error) {
      throw translate(error);
    }
  }

  async signOut(): Promise<void> {
    try {
      await signOut(getFirebaseClientAuth());
    } catch (error) {
      throw translate(error);
    }
  }
}

File: apps/mobile/lib/auth/firebase_gateway.dart
import 'package:firebase_auth/firebase_auth.dart';

const _authActionUrl = String.fromEnvironment('AUTH_ACTION_URL');

class ClientAuthException implements Exception {
  const ClientAuthException(this.code, this.message);

  final String code;
  final String message;
}

class AuthSession {
  const AuthSession({
    required this.uid,
    required this.email,
    required this.emailVerified,
  });

  final String uid;
  final String? email;
  final bool emailVerified;
}

abstract interface class AuthGateway {
  Stream<AuthSession?> authStateChanges();
  Future<String?> getIdToken();
  Future<AuthSession> refreshSession();
  Future<void> sendEmailVerification();
  Future<void> sendPasswordResetEmail(String email);
  Future<AuthSession> signIn(String email, String password);
  Future<void> signOut();
}

class FirebaseAuthGateway implements AuthGateway {
  FirebaseAuthGateway({FirebaseAuth? auth})
      : _auth = auth ?? FirebaseAuth.instance;

  final FirebaseAuth _auth;

  static AuthSession _session(User user) => AuthSession(
        uid: user.uid,
        email: user.email,
        emailVerified: user.emailVerified,
      );

  @override
  Stream<AuthSession?> authStateChanges() => _auth.authStateChanges().map(
        (user) => user == null ? null : _session(user),
      );

  @override
  Future<String?> getIdToken() async {
    final user = _auth.currentUser;
    return user == null ? null : await user.getIdToken();
  }

  @override
  Future<AuthSession> signIn(String email, String password) async {
    try {
      final credential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      final user = credential.user;
      if (user == null) {
        throw const ClientAuthException(
          'client-error',
          'Firebase returned no authenticated user',
        );
      }
      return _session(user);
    } on FirebaseAuthException catch (error) {
      throw ClientAuthException(
        error.code,
        error.message ?? 'Authentication failed',
      );
    }
  }

  @override
  Future<AuthSession> refreshSession() async {
    final user = _auth.currentUser;
    if (user == null) {
      throw const ClientAuthException(
        'no-current-user',
        'No authenticated user is available',
      );
    }
    try {
      await user.reload();
      final refreshed = _auth.currentUser;
      if (refreshed == null) {
        throw const ClientAuthException(
          'no-current-user',
          'No authenticated user is available',
        );
      }
      await refreshed.getIdToken(true);
      return _session(refreshed);
    } on FirebaseAuthException catch (error) {
      throw ClientAuthException(
        error.code,
        error.message ?? 'Unable to refresh the session',
      );
    }
  }

  @override
  Future<void> sendEmailVerification() async {
    final user = _auth.currentUser;
    if (user == null) {
      throw const ClientAuthException(
        'no-current-user',
        'No authenticated user is available',
      );
    }
    try {
      await user.sendEmailVerification();
    } on FirebaseAuthException catch (error) {
      throw ClientAuthException(
        error.code,
        error.message ?? 'Unable to send the verification email',
      );
    }
  }

  @override
  Future<void> sendPasswordResetEmail(String email) async {
    try {
      await _auth.sendPasswordResetEmail(
        email: email,
        actionCodeSettings: _authActionUrl.isEmpty
            ? null
            : ActionCodeSettings(
                url: _authActionUrl,
                handleCodeInApp: false,
              ),
      );
    } on FirebaseAuthException catch (error) {
      throw ClientAuthException(
        error.code,
        error.message ?? 'Unable to send the reset email',
      );
    }
  }

  @override
  Future<void> signOut() async {
    try {
      await _auth.signOut();
    } on FirebaseAuthException catch (error) {
      throw ClientAuthException(error.code, error.message ?? 'Sign out failed');
    }
  }
}
